### PR TITLE
Feat/captcha

### PR DESCRIPTION
**Pull request by Lucas-Bento-D**
PR about issue: https://github.com/detrash/recy-network/issues/170

### CODE REVIEW

**Comment by Lucas-Bento-D** [comment]
@coderabbitai review

File: src/modules/cloudflare/captcha/captcha.service.ts [hunk 1]
```diff
@@ -0,0 +1,26 @@
+import { Injectable, InternalServerErrorException } from '@nestjs/common';
+
+import { TokenSchemaDto } from './dtos/token.dto';
+
+@Injectable()
+export class CaptchaService {
+  async captchaVerification({ token }: TokenSchemaDto): Promise<any> {
```
Comment (yurimutti) [inline src/modules/cloudflare/captcha/captcha.service.ts]: Coud you avoid this any and use correct type ?

File: src/modules/cloudflare/captcha/captcha.controller.ts [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+import { CaptchaService } from './captcha.service';
+import { TokenSchema, TokenSchemaDto } from './dtos/token.dto';
+@ApiTags('captcha')
+@Controller({ path: 'captcha', version: '1' })
+export class CaptchaController {
+  constructor(private readonly captchaService: CaptchaService) {}
+
+  @Post()
+  @UsePipes(new ZodValidationPipe(TokenSchema))
+  async captcha(@Body() TokenSchemaDto: TokenSchemaDto): Promise<any> {
```
Comment (yurimutti) [inline src/modules/cloudflare/captcha/captcha.controller.ts]: Coud you avoid this any and use correct type ?

File: src/modules/cloudflare/captcha/captcha.service.ts [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+
+      const result = await fetch(url, {
+        body: formData,
+        method: 'POST',
+      });
+
+      const data = await result.json();
+      return data;
+    } catch (error) {
+      throw new InternalServerErrorException();
```
Comment (yurimutti) [inline src/modules/cloudflare/captcha/captcha.service.ts]: Could update to `throw new InternalServerErrorException('Failed turnstile verification.')` is more descriptive.

File: src/modules/cloudflare/captcha/captcha.service.ts (hunk 1 above)
Comment (yurimutti) [reply src/modules/cloudflare/captcha/captcha.service.ts]: Example: 
```
interface CaptchaVerificationResponse {
  success: boolean;
  challenge_ts?: string;
  hostname?: string;
  'error-codes'?: string[];
}

...Promise<CaptchaVerificationResponse>
```

File: src/modules/cloudflare/captcha/captcha.service.ts [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+
+import { TokenSchemaDto } from './dtos/token.dto';
+
+@Injectable()
+export class CaptchaService {
+  async captchaVerification({ token }: TokenSchemaDto): Promise<any> {
+    try {
+      const formData = new FormData();
+      formData.append('secret', process.env.CAPTCHA_SECRET_KEY);
+      formData.append('response', token);
```
Comment (yurimutti) [inline src/modules/cloudflare/captcha/captcha.service.ts]: could you moving please to configService ?

```
const secretKey = this.configService.get<string>('CAPTCHA_SECRET_KEY');
const hostVerificationUrl = this.configService.get<string>('CAPTCHA_HOST_VERIFICATION');

if (!secretKey || !hostVerificationUrl) {
  throw new Error('CAPTCHA environment variables are not configured properly.');
}
```

Comment (yurimutti) [reply src/modules/cloudflare/captcha/captcha.service.ts]: https://docs.nestjs.com/techniques/configuration

File: src/modules/cloudflare/captcha/captcha.service.ts (hunk 3 above)
Comment (yurimutti) [reply src/modules/cloudflare/captcha/captcha.service.ts]: Could you create a comment for implement logger soon ?

`// TODO:  Log the detailed error for internal debugging`

File: src/modules/cloudflare/captcha/captcha.service.ts [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+
+      const url = `${process.env.CAPTCHA_HOST_VERIFICATION}/turnstile/v0/siteverify`;
+
+      const result = await fetch(url, {
+        body: formData,
+        method: 'POST',
+      });
+
+      const data = await result.json();
+      return data;
```
Comment (yurimutti) [inline src/modules/cloudflare/captcha/captcha.service.ts]: We need to check some errors here:

Maybe this work we need to check the doc:


```
 if (!result.ok) {
        throw new Error(`CAPTCHA verification failed with status: ${result.status}`);
      }

      const data: CaptchaVerificationResponse = await result.json();

      if (!data.success) {
        throw new Error(`CAPTCHA verification error: ${data['error-codes']?.join(', ') || 'Unknown error'}`);
      }

```

File: src/modules/cloudflare/captcha/captcha.controller.ts [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+    return await this.captchaService.captchaVerification(TokenSchemaDto);
+  }
+}
```
Comment (yurimutti) [inline src/modules/cloudflare/captcha/captcha.controller.ts]: We need be more flexible here and use a guard.

**Comment by yurimutti** [comment]
Instead of having a dedicated controller for captcha verification, we propose using a guard (TurnstileGuard) to check the captcha token in the request before accessing protected routes.

The guard will handle the following:

Validate the captcha token received in the request.
Return BadRequestException for missing or invalid tokens.
Only allow access to the route if captcha verification is successful.

**TurnstileGuard:**

```
export enum Messages {
  FAIL = 'Failed turnstile verification.',
  MISSING = 'Missing turnstile verification code.',
  INVALID = 'Invalid turnstile verification code.',
}

@Injectable()
export class TurnstileGuard implements CanActivate {
  constructor(
    private readonly turnstileService: TurnstileService,
    @Inject('TurnstileServiceOptions')
    private readonly options: ITurnstileOptions,
  ) {}

  async canActivate(context: ExecutionContext): Promise<boolean> {
    const request = context.switchToHttp().getRequest();
    const responseToken = this.options.tokenResponse(request);

    // If token is missing, throw error
    if (!responseToken) throw new BadRequestException(Messages.MISSING);

    // Validate the token using Turnstile service
    const { success } = await this.turnstileService.validateToken(responseToken);

    // If validation fails, throw error
    if (!success) throw new BadRequestException(Messages.INVALID);

    return success;
  }
}
```

**Adapting the TurnstileService:**
import { Injectable, Inject } from '@nestjs/common';
import { TurnstileOptions } from './interfaces/turnstile-options.interface';

```
@Injectable()
export class TurnstileService {
  constructor(
    @Inject('TurnstileServiceOptions')
    private readonly options: TurnstileOptions,
  ) {}

  async validateToken(token: string): Promise<{ success: boolean; error?: string }> {
    // Logic to validate the Turnstile token
    // For example, make a call to Turnstile API or perform custom validation here
    
    // Sample return (you would call the actual Turnstile service here):
    const validationResponse = await someExternalApiCall(token, this.options.secretKey);
    
    if (validationResponse.success) {
      return { success: true };
    } else {
      return { success: false, error: 'Invalid token' };
    }
  }
}

```

**Updating the CaptchaService:**
```
import { Injectable, InternalServerErrorException, Inject } from '@nestjs/common';
import { TokenSchemaDto } from './dtos/token.dto';
import { TurnstileService } from './turnstile.service';
import { ITurnstileOptions } from './interfaces/turnstile-options.interface';

@Injectable()
export class CaptchaService {
  constructor(
    private readonly turnstileService: TurnstileService,
    @Inject('TurnstileServiceOptions')
    private readonly options: ITurnstileOptions,
  ) {}

  async captchaVerification({ token }: TokenSchemaDto): Promise<any> {
    try {
      // Using the TurnstileService to validate the captcha token
      const { success, error } = await this.turnstileService.validateToken(token);

      // If validation fails, throw an exception
      if (!success) {
        throw new InternalServerErrorException(error || 'Captcha verification failed');
      }

      return { success: true, message: 'Captcha validated successfully' };
    } catch (error) {
      // Catch any errors and throw InternalServerErrorException
      throw new InternalServerErrorException(error.message || 'Internal server error during captcha verification');
    }
  }
}

```

TurnstileGuard: A guard to validate captcha tokens before accessing routes.
TurnstileService: A service responsible for validating the Turnstile token using an external service or custom validation logic.
CaptchaService: A service that delegates captcha verification to TurnstileService and handles errors appropriately.

**Example**
```

import { Controller, Get, Post, UseGuards, Body } from '@nestjs/common';
import { TurnstileGuard } from './guards/turnstile.guard';
import { CaptchaService } from './services/captcha.service';
import { TokenSchemaDto } from './dtos/token.dto';

@Controller('example')
export class ExampleController {
  constructor(private readonly captchaService: CaptchaService) {}

  // Public route (no captcha validation required)
  @Get('public')
  publicRoute() {
    return { message: 'This is a public endpoint. No captcha required.' };
  }

  // Protected route (captcha validation required)
  @UseGuards(TurnstileGuard)
  @Post('protected')
  async protectedRoute(@Body() body: TokenSchemaDto) {
    return { message: 'Captcha validation successful! You accessed a protected route.', data: body };
  }

  // Another example using the CaptchaService directly
  @Post('validate-captcha')
  async validateCaptcha(@Body() body: TokenSchemaDto) {
    const result = await this.captchaService.captchaVerification(body);
    return {
      message: 'Captcha verification completed.',
      success: result.success,
    };
  }
}

```

**Review by yurimutti — CHANGES_REQUESTED** [review]
- Create TurnstileGuard
- Create TurnstileService
- Create CaptchaService

**Comment by yurimutti** [comment]
https://docs.nestjs.com/guards

File: src/modules/cloudflare/captcha/captcha.controller.ts [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+
+  @Post('validate-captcha')
+  async validateCaptcha(@Body() body: TokenSchemaDto): Promise<Omit<ICaptchaProtectedReturn, 'data'>> {
+    const result = await this.captchaService.captchaVerification(body);
+    return {
+      message: 'Captcha verification completed.',
+      success: result.success,
+    };
+  }
+}
```
Comment (yurimutti) [inline src/modules/cloudflare/captcha/captcha.controller.ts]: We dont need this controller we will use the GUARD on the route to be protected.

Comment (yurimutti) [reply src/modules/cloudflare/captcha/captcha.controller.ts]: Example of use

On users controller we put the captcha guard and the frontend need to send the captcha token.

Comment (yurimutti) [reply src/modules/cloudflare/captcha/captcha.controller.ts]: image

File: src/modules/cloudflare/turnstile/types.ts [hunk 8]
```diff
@@ -0,0 +1,7 @@
+export interface ITurnstileVerificationResponse {
+  success: boolean;
+  challenge_ts?: string;
+  hostname?: string;
+  'error-codes'?: string[];
+  metadata: { interactive: boolean };
+}
```
Comment (yurimutti) [inline src/modules/cloudflare/turnstile/types.ts]: We didint use the prefix I on the project, please could you remove this ? -> TurnstileVerificationResponse

File: src/modules/cloudflare/captcha/types.ts [hunk 9]
```diff
@@ -0,0 +1,12 @@
...
+    message: string,
+    data: {
+        token: string
+    }
+}
+export interface ITokenResponse{
+    body: {
+        token: string
+    }
+}
```
Comment (yurimutti) [inline src/modules/cloudflare/captcha/types.ts]: We didint use the prefix I on the project, please could you remove this ?

File: src/modules/cloudflare/turnstile/turnstile.service.ts [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+
+    if (!data.success) {
+      throw new Error(
+        `CAPTCHA verification error: ${
+          data['error-codes']?.join(', ') || 'Unknown error'
+        }`,
+      );
+    }
+
+    return data;
```
Comment (yurimutti) [inline src/modules/cloudflare/turnstile/turnstile.service.ts]: Its interesting use a try catch here and use nestjs exceptions. throw new BadRequestException('...');

  throw new BadRequestException()

```
try {
      const result = await fetch(url, {
        body: formData,
        method: 'POST',
      });

      if (!result.ok) {
        throw new BadRequestException(
          `CAPTCHA verification failed with status: ${result.status}`,
        );
      }

      const data = (await result.json()) as ITurnstileVerificationResponse;

      if (!data.success) {
        throw new UnauthorizedException(
          `CAPTCHA verification error: ${
            data['error-codes']?.join(', ') || 'Unknown error'
          }`,
        );
      }

      return data;
    } catch (error) {
      throw new InternalServerErrorException(
        `Error occurred during CAPTCHA verification: ${error.message}`,
      );
    }
  }
```

File: src/modules/cloudflare/turnstile/turnstile.service.ts [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+    token: string,
+    secretKey: string,
+  ): Promise<ITurnstileVerificationResponse> {
+    const formData = new FormData();
+    const hostVerificationUrl = this.options.host;
+
+    if (!secretKey || !hostVerificationUrl) {
+      throw new Error(
+        'CAPTCHA environment variables are not configured properly.',
+      );
```
Comment (yurimutti) [inline src/modules/cloudflare/turnstile/turnstile.service.ts]:
```
InternalServerErrorException(
        `CAPTCHA environment variables are not configured properly. Secret Key: ${secretKey} Host Verification Url:${hostVerificationUrl}`,
      );
```

File: src/modules/cloudflare/captcha/captcha.service.ts [hunk 12]
```diff
@@ -0,0 +1,30 @@
...
+      const { success, error } = await this.turnstileService.validateToken(token);
+
+      if (!success) {
+        throw new InternalServerErrorException(error || 'Captcha verification failed');
+      }
+
+      return { success: true, message: 'Captcha validated successfully' };
+    } catch (error) {
+      // TODO:  Log the detailed error for internal debugging
+      throw new InternalServerErrorException('Failed turnstile verification.');
```
Comment (yurimutti) [inline src/modules/cloudflare/captcha/captcha.service.ts]: `throw new InternalServerErrorException(`Captcha verification failed ${error}` );
`
more context.

Comment (yurimutti) [reply src/modules/cloudflare/captcha/captcha.service.ts]: throw new InternalServerErrorException(`'Failed turnstile verification. ${error}`);

**Review by yurimutti — CHANGES_REQUESTED** [review]
- Remove **/captcha.controller.ts** we will use guard on endpoints 
- Replace throw new Error() with specific exception classes, such as BadRequestException and InternalServerErrorException.
- Add Error Handling with Try/Catch
- More Structured Validation
- Remove I to named the interfaces (we didin't use this pattern)

**Comment by yurimutti** [comment]
`'X-recaptcha-token': recaptchaToken`

File: src/modules/cloudflare/captcha/captcha.controller.ts [hunk 13]
```diff
@@ -0,0 +1,30 @@
...
+import { TurnstileGuard } from '../turnstile/turnstile.guard';
+@ApiTags('captcha')
+@Controller({ path: 'captcha', version: '1' })
+export class CaptchaController {
+  constructor(private readonly captchaService: CaptchaService) {}
+  
+  @UseGuards(TurnstileGuard)
+  @Post('protected')
+  @UsePipes(new ZodValidationPipe(TokenSchema))
+  async captchaProtected(@Body() body: TokenSchemaDto): Promise<ICaptchaProtectedReturn> {
```
Comment (bernardinorafael) [inline src/modules/cloudflare/captcha/captcha.controller.ts]: Please, let's remove the `I` prefix on `types/interfaces`

File: src/modules/cloudflare/captcha/captcha.controller.ts [hunk 14]
```diff
@@ -0,0 +1,30 @@
...
+    return { success: true, message: 'Captcha validation successful! You accessed a protected route.', data: body };
```
Comment (bernardinorafael) [inline src/modules/cloudflare/captcha/captcha.controller.ts]: Messages that comes from the API is not for user reading, we can't put the API messages directly on toasts...

File: src/modules/cloudflare/turnstile/turnstile.guard.ts [hunk 15]
```diff
@@ -0,0 +1,39 @@
...
+  CanActivate,
+  ExecutionContext,
+  Inject,
+  Injectable,
+} from '@nestjs/common';
+
+import { TurnstileOptions } from './interfaces/turnstile-options.interface';
+import { TurnstileService } from './turnstile.service';
+
+export enum Messages {
```
Comment (bernardinorafael) [inline src/modules/cloudflare/turnstile/turnstile.guard.ts]: This is a personal guess, we don't need this `enum`, it is being initialized too close to where it is used, this adds unnecessary complexity in a small block of code...

File: src/modules/cloudflare/turnstile/types.ts [hunk 16]
```diff
@@ -0,0 +1,7 @@
+export interface ITurnstileVerificationResponse {
```
Comment (bernardinorafael) [inline src/modules/cloudflare/turnstile/types.ts]: I believe we should not use `snake_case` and `kebab-case` in JS/TS code at the same time, let's use `camelCase` instead

**Review by bernardinorafael — CHANGES_REQUESTED** [review]

File: src/modules/cloudflare/turnstile/types.ts (hunk 16 above)
Comment (yurimutti) [reply src/modules/cloudflare/turnstile/types.ts]: Good point! camelCase and in same cases Pascal Case.

File: src/modules/cloudflare/turnstile/turnstile.guard.ts (hunk 15 above)
Comment (yurimutti) [reply src/modules/cloudflare/turnstile/turnstile.guard.ts]: Maybe make sense move to a interface file

File: src/modules/cloudflare/captcha/captcha.controller.ts (hunk 7 above)
Comment (Lucas-Bento-D) [reply src/modules/cloudflare/captcha/captcha.controller.ts]: controller removed

File: src/modules/cloudflare/turnstile/types.ts (hunk 8 above)
Comment (Lucas-Bento-D) [reply src/modules/cloudflare/turnstile/types.ts]: prefix removed

File: src/modules/cloudflare/captcha/types.ts (hunk 9 above)
Comment (Lucas-Bento-D) [reply src/modules/cloudflare/captcha/types.ts]: prefix removed

File: src/modules/cloudflare/turnstile/turnstile.service.ts (hunk 10 above)
Comment (Lucas-Bento-D) [reply src/modules/cloudflare/turnstile/turnstile.service.ts]: service improvemented

File: src/modules/cloudflare/turnstile/turnstile.service.ts (hunk 11 above)
Comment (Lucas-Bento-D) [reply src/modules/cloudflare/turnstile/turnstile.service.ts]: treatment of error improvemented

File: src/modules/cloudflare/captcha/captcha.controller.ts (hunk 13 above)
Comment (Lucas-Bento-D) [reply src/modules/cloudflare/captcha/captcha.controller.ts]: with guard uses, we will to remove captcha.controller, but prefix I was removed of all types

File: src/modules/cloudflare/turnstile/types.ts (hunk 16 above)
Comment (Lucas-Bento-D) [reply src/modules/cloudflare/turnstile/types.ts]: The use of type follows return of cloudflare about turnstile, With successfully or not, will return how was structured on type.

File: src/modules/cloudflare/turnstile/turnstile.guard.ts (hunk 15 above)
Comment (Lucas-Bento-D) [reply src/modules/cloudflare/turnstile/turnstile.guard.ts]: Yeah, I'm agree about your point.

File: src/modules/cloudflare/captcha/captcha.controller.ts (hunk 14 above)
Comment (Lucas-Bento-D) [reply src/modules/cloudflare/captcha/captcha.controller.ts]: with guard uses, we will to remove captcha.controller

**Comment by Lucas-Bento-D** [comment]
The captcha module removal happened because Turnstile guard flux.
while I testing, I saw that for using the guard, we would need to inject, Turnstile service and providers, but nothing about captcha module, so I created a specific module for turnstile exporting service and provider and deleted the captcha structure. Therewith in turnstile guard use, we just need to import TurnstileModule, facilitating maintenance.

**Comment by Lucas-Bento-D** [comment]
> `'X-recaptcha-token': recaptchaToken`

**Review by yurimutti — APPROVED** [review]
LGFM

**Pull request merged**